src/Compilers/Core/Portable/Text/CompositeText.cs

```diff
@@ -160,16 +166,214 @@ public override void CopyTo(int sourceIndex, char[] destination, int destination
+        /// the minimum segment size are combined together.
+        /// </summary>
+        private static int GetReducedSize(ArrayBuilder<SourceText> segments, int minimumSegmentSize)
+        {
+            int numberOfSegmentsReduced = 0;
+
+            for (int i = 0; i < segments.Count - 1; i++)
+            {
+                if (segments[i].Length <= minimumSegmentSize)
+                {
+                    // count how many contiguous segments are reducible
+                    int count = 1;
+                    for (int j = i + 1; j < segments.Count; j++)
+                    {
+                        if (segments[j].Length <= minimumSegmentSize)
+                        {
+                            count++;
+                        }
+                    }
+
+                    if (count > 1)
+                    {
+                        var removed = count - 1;
+                        numberOfSegmentsReduced += removed;
+                        i += removed;
+                    }
+                }
+            }
+
+            return segments.Count - numberOfSegmentsReduced;
+        }
+
+        private static ObjectPool<HashSet<SourceText>> _uniqueSourcesPool
+            = new ObjectPool<HashSet<SourceText>>(() => new HashSet<SourceText>(), 10);
+
+        /// <summary>
+        /// Compute total text length and total size of storage buffers held
+        /// </summary>
+        private static void ComputeLengthAndSize(IReadOnlyList<SourceText> segments, out int length, out int size)
+        {
+            var uniqueSources = _uniqueSourcesPool.Allocate();
+
+            length = 0;
+            for (int i = 0; i < segments.Count; i++)
+            {
+                var segment = segments[i];
+                length += segment.Length;
+                uniqueSources.Add(segment.StorageKey);
+            }
+
+            size = 0;
+            foreach (var segment in uniqueSources)
+            {
+                size += segment.Size;
+            }
+
+            uniqueSources.Clear();
+            _uniqueSourcesPool.Free(uniqueSources);
+        }
+
+        /// <summary>
+        /// Trim excessive inaccessible text.
+        /// </summary>
+        private static void TrimInaccessibleText(ArrayBuilder<SourceText> segments)
+        {
+            int length, size;
+            ComputeLengthAndSize(segments, out length, out size);
+
+            // if more than half of the storage is unused, compress into a single new segment
+            if (length < size / 2)
+            {
+                var encoding = segments[0].Encoding;
+                var algorithm = segments[0].ChecksumAlgorithm;
+
+                var writer = SourceTextWriter.Create(encoding, algorithm, length);
+                foreach (var segment in segments)
+                {
+                    segment.Write(writer);
+                }
+
+                segments.Clear();
```

Comment:
It's not obvious to me from the caller that this is mutating ArrayBuilder. Updating the doc comment might help a bit.